src/org/akvo/resumed.clj

```diff
@@ -120,26 +120,52 @@
             (DatatypeConverter/parseBase64Binary)
             (String.))))))
 
+(defn get-host
+  "Returns the HOST for a given request
+  It attempts to honor: X-Forwared-Host, Origin, Host headers"
+  [req]
+  (let [host (or (get-header req "x-forwarded-host")
+                 (get-header req "origin")
+                 (first (.split (or (get-header req "host") "") ":")))]
+    (if (str/blank? host)
+      (:server-name req)
+      host)))
+
+(defn get-protocol
+  "Returns the protocol #{\"http\" \"https\"} for a given request"
+  [req]
+  (let [forwarded-proto (get-header req "x-forwarded-proto")
+        known-protocols #{"http" "https"}]
+    (if (or (str/blank? forwarded-proto)
+            (not (known-protocols forwarded-proto)))
+      (name (:scheme req))
+      forwarded-proto)))
+
+(def ^:const http-default-ports #{443 80})
+
+(defn get-port
+  "Returns the port of the request,
+  Empty if port is 80, 443 as those are default ports
+  Empty for forwared requests (server behind a proxy)"
+  [req]
+  (let [forwarded (or (get-header req "x-forwarded-host")
+                      (get-header req "x-forwarded-proto"))
+        host (get-header req "host")]
```

Comment:
move this `host` to an if-let in line 157, so the context of the host is smaller (I do not need to parse it in the first part of the if condition), plus we remove the and in line 157.